[REF] ProfileConverter

SamuilsRulovs opened the pull request:
Refactoring  "ProfileConverter.h". 

Added Functions:  
- AddRectangleCoordiantes 
- AddTrapeziumCoordinates 
- addArc_or_push.back 
- SearchExistingVertex
- convertIfcOpenCrossProfileDef

Refactored Functions:
- "convertIfcParametrizedProfileDef" (Extracted SUBTYPES, added Doxy-comments, replaced "shared_ptr" with "EXPRESSReference")
- "computeProfile" (replaced "shared_ptr" with "EXPRESSReference")
- "convertIfcArbitraryOpenProfileDef"(replaced "shared_ptr" with "EXPRESSReference")
- "convertIfcCompositeProfileDef"(replaced "shared_ptr" with "EXPRESSReference")
- "convertIfcDerivedProfileDef"(replaced "shared_ptr" with "EXPRESSReference")
- "convertIfcParameterizedProfileDefWithPosition"(replaced "shared_ptr" with "EXPRESSReference")

Commit:
Corrected "IfcParametrizedProfileDef"(Added Functions GetRectangleCoordiantes,GetTrapeziumCoordinates, GetCircleRadius, GetAsymmetricIShapeCoordinates)

pjanck commented on Core/src/IfcGeometryConverter/ProfileConverter.h:
```diff
@@ -539,37 +539,16 @@ namespace OpenInfraPlatform {
 								if(hollow->InnerFilletRadius) {
 									r2 = hollow->InnerFilletRadius * length_factor;
 								}
```
```suggestion
								double r2 = hollow->InnerFilletRadius.value_or(0.0) * length_factor;
```
This is how you can handle Optional<> lengths, angles, and so on. It condenses the code and makes it more readable.

pjanck commented on Core/src/IfcGeometryConverter/ProfileConverter.h:
```diff
@@ -1522,6 +1477,64 @@ namespace OpenInfraPlatform {
 
+			// Function 10: Get rectangle coordiantes
+			
+			void GetRectangleCoordinates (std::vector<carve::geom::vector<2>>& coords, double radius, double x, double y) const
```
It is best, if input parameters are marked as const:
```suggestion
			void GetRectangleCoordinates (std::vector<carve::geom::vector<2>>& coords, const double radius, const double x, const double y) const

```

pjanck commented on Core/src/IfcGeometryConverter/ProfileConverter.h:
```diff
@@ -1522,6 +1477,64 @@ namespace OpenInfraPlatform {
 
+			// Function 10: Get rectangle coordiantes
+			
+			void GetRectangleCoordinates (std::vector<carve::geom::vector<2>>& coords, double radius, double x, double y) const
```
Also, about naming: the function *adds* coordinates to `coords` - I would call it AddRectangleCoordinates.
```suggestion
			void AddRectangleCoordinates (std::vector<carve::geom::vector<2>>& coords, double radius, double x, double y) const

```

SamuilsRulovs replied:
Done

pjanck commented on Core/src/IfcGeometryConverter/ProfileConverter.h:
```diff
@@ -1522,6 +1477,64 @@ namespace OpenInfraPlatform {
+		
+
+			//Function 11: GetTrapeziumCoordinates
+			void GetTrapeziumCoordinates(std::vector<carve::geom::vector<2>>& coords, double xBottom, double xOffset, double xTop, double y)const
```
Similar comments as for GetRectangleCoordinates.

SamuilsRulovs replied:
Done

pjanck commented on Core/src/IfcGeometryConverter/ProfileConverter.h:
```diff
@@ -1522,6 +1477,64 @@ namespace OpenInfraPlatform {
+			}
+
+			//Function 12: GetCircleRadius
+			void GetCircleCoordinates(std::vector<carve::geom::vector<2>>& coords, double radius, double d_angle, int num_segments ) const
```
Similar comments as for GetRectangleCoordinates.

SamuilsRulovs replied:
Done

pjanck commented on Core/src/IfcGeometryConverter/ProfileConverter.h:
```diff
@@ -1522,6 +1477,64 @@ namespace OpenInfraPlatform {
+			void GetCircleCoordinates(std::vector<carve::geom::vector<2>>& coords, double radius, double d_angle, int num_segments ) const
+			{
+				
+				double angle = 0;
```
```suggestion
				double angle = 0.;
```
Always initialize `double`-s with a floating point number, not an integer.

SamuilsRulovs replied:
Done

pjanck commented on Core/src/IfcGeometryConverter/ProfileConverter.h:
```diff
@@ -1522,6 +1477,64 @@ namespace OpenInfraPlatform {
+			}
+
+			//Function 12: GetCircleRadius
+			void GetCircleCoordinates(std::vector<carve::geom::vector<2>>& coords, double radius, double d_angle, int num_segments ) const
```
Please have a look at the `addArc` function - you can call it and get what you need.

SamuilsRulovs replied:
Done

pjanck commented on Core/src/IfcGeometryConverter/ProfileConverter.h:
```diff
@@ -539,37 +539,16 @@ namespace OpenInfraPlatform {
+							
 								// Inner
 								std::vector<carve::geom::vector<2>> inner_loop;
 								x -= 2 * t;
```
In order to not confuse compiler with implicit casting between int and double, use doubles in calculations with doubles.
```suggestion
								x -= 2. * t;
```

SamuilsRulovs replied:
Done

pjanck commented on Core/src/IfcGeometryConverter/ProfileConverter.h:
```diff
@@ -539,37 +539,16 @@ namespace OpenInfraPlatform {
 								// Inner
 								std::vector<carve::geom::vector<2>> inner_loop;
 								x -= 2 * t;
 								y -= 2 * t;
```
Similarly
```suggestion
								y -= 2. * t;
```

SamuilsRulovs replied:
Done

pjanck commented on Core/src/IfcGeometryConverter/ProfileConverter.h:
```diff
@@ -582,20 +561,16 @@ namespace OpenInfraPlatform {
-								addArc(outer_loop, rr, M_PI_2, M_PI_2, -x * 0.5 + rr, y * 0.5 - rr);
-								addArc(outer_loop, rr, M_PI, M_PI_2, -x * 0.5 + rr, -y * 0.5 + rr);
-								addArc(outer_loop, rr, 3 * M_PI_2, M_PI_2, x * 0.5 - rr, -y * 0.5 + rr);
+								GetRectangleCoordinates(outer_loop, rr, x, y);
```
Even nicer:
```suggestion
								GetRectangleCoordinates(outer_loop, rounded_rectangle->RoundingRadius.value_or(0.), x, y);
```

SamuilsRulovs replied:
Done

pjanck commented on Core/src/IfcGeometryConverter/ProfileConverter.h:
```diff
@@ -625,29 +598,29 @@ namespace OpenInfraPlatform {
 					if(radius < 0.000001) {
```
```suggestion
					if(radius < GeomSettings()->getPrecision()) {
```
We now have a function that tells what precision a file guaranties.

SamuilsRulovs replied:
Done

pjanck commented on Core/src/IfcGeometryConverter/ProfileConverter.h:
```diff
@@ -625,29 +598,29 @@ namespace OpenInfraPlatform {
-						angle += d_angle;
-					}
+					//Function GetCircleCoordinates
+					GetCircleCoordinates(outer_loop, radius, d_angle, num_segments);
```
You can use the already present function `addArc`

SamuilsRulovs replied:
Done

pjanck commented on Core/src/IfcGeometryConverter/ProfileConverter.h:
```diff
@@ -625,29 +598,29 @@ namespace OpenInfraPlatform {
-							angle += d_angle2;
-						}
+						//Function GetCircleCoordinates
+						GetCircleCoordinates(inner_loop, radius2, d_angle2, num_segments2);
```
Similar to above, use `addArc`

SamuilsRulovs replied:
Done

pjanck commented on Core/src/IfcGeometryConverter/ProfileConverter.h:
```diff
@@ -701,23 +674,12 @@ namespace OpenInfraPlatform {
-								else {
-									outer_loop.push_back(carve::geom::VECTOR(tw * 0.5, (h * 0.5 - tfTop)));
-								}
+								double bTop = asym_I_profile->TopFlangeWidth * length_factor;																
+								double tfTop = asym_I_profile->TopFlangeThickness.value_or(tf);
+								double rTop = asym_I_profile->TopFlangeFilletRadius.value_or(r);
```
Awesome!

SamuilsRulovs replied:
Thank you)

pjanck commented on Core/src/IfcGeometryConverter/ProfileConverter.h:
```diff
@@ -701,23 +674,12 @@ namespace OpenInfraPlatform {
+								double tfTop = asym_I_profile->TopFlangeThickness.value_or(tf);
+								double rTop = asym_I_profile->TopFlangeFilletRadius.value_or(r);
+								//Function GetAsymmetricCoordinates
+								GetAsymmetricIShapeCoordinates(outer_loop, rTop, tw, tfTop, h);
```
I don't really like this function. It only adds one corner of the I-Shape and not hte whole as its name conveys. Please leave this functionality currently where it is (so, remove the function and shift the if-statement back). The whole handling of the I-Shape will get its own function.

SamuilsRulovs replied:
Done

pjanck commented on Core/src/IfcGeometryConverter/ProfileConverter.h:
```diff
@@ -739,39 +701,27 @@ namespace OpenInfraPlatform {
-							b = l_shape->Width * length_factor;
-						}
-
+						
+						double b = l_shape->Width.value_or(h);
+						
```
A comment on the general style: no need for unnecessary line breaks. Usually, line breaks come as a new paragraph - so, if you are only initializing some variables, this is seen as a continuous paragraph, as the same sort of actions are done.

SamuilsRulovs replied:
Done

pjanck commented on Core/src/IfcGeometryConverter/ProfileConverter.h:
```diff
@@ -1522,6 +1477,64 @@ namespace OpenInfraPlatform {
+					coords.push_back(carve::geom::VECTOR((radius * cos(angle)), (radius * sin(angle))));
+					angle += d_angle;
+				}
+				
```
For example, also this line break is unnecessary.

SamuilsRulovs replied:
Done

pjanck commented on Core/src/IfcGeometryConverter/ProfileConverter.h:
```diff
@@ -1522,6 +1477,64 @@ namespace OpenInfraPlatform {
 
+			// Function 10: Get rectangle coordiantes
+			
+			void GetRectangleCoordinates (std::vector<carve::geom::vector<2>>& coords, double radius, double x, double y) const
```
Please add Doxy-comments to this function. It seems done.

SamuilsRulovs replied:
In progress

SamuilsRulovs replied:
Done

pjanck reviewed (changes requested):
Good progress in the correct general direction. Great functions `GetRectangleCoordinates` and `GetTrapezoidCoordinates`. Both should get their doxy-comments.

Commit:
Merge remote-tracking branch 'tumcms/development' into RefactoringIfcRectangleProfileDef

SamuilsRulovs committed:
Update GetRectangleCoordinates (marked input parameters as const)

Co-authored-by: Stefan Jaud <59165496+pjanck@users.noreply.github.com>

SamuilsRulovs committed:
remade if special function with value_or function

Co-authored-by: Stefan Jaud <59165496+pjanck@users.noreply.github.com>

hechth commented:
@SamuilsRulovs Please reply to all comments to point out what is resolved and what not!

Commit:
Corrected Functions in IfcProfileDef (AddCircleCoordinates, AddRectangleCoordinates, AddTrapeziumCoordinates)

pjanck commented on Core/src/IfcGeometryConverter/ProfileConverter.h:
```diff
@@ -530,24 +530,16 @@ namespace OpenInfraPlatform {
+								double r2 = hollow->InnerFilletRadius.value_or(0.0) * length_factor;
 								// Outer
-								GetRectangleCoordinates(outer_loop, r1, x, y);
+								AddRectangleCoordinates(outer_loop, r1, x, y);
```
Super!

SamuilsRulovs replied:
Thank you)

pjanck commented on Core/src/IfcGeometryConverter/ProfileConverter.h:
```diff
@@ -675,10 +665,10 @@ namespace OpenInfraPlatform {
 								double bTop = asym_I_profile->TopFlangeWidth * length_factor;																
-								double tfTop = asym_I_profile->TopFlangeThickness.value_or(tf);
-								double rTop = asym_I_profile->TopFlangeFilletRadius.value_or(r);
+								double tfTop = asym_I_profile->TopFlangeThickness.value_or(tf/ length_factor) * length_factor;
+								double rTop = asym_I_profile->TopFlangeFilletRadius.value_or(r/ length_factor) * length_factor;
```
Use direct references to the default value - i.e. i_shape->FlangeThickness for tf.

SamuilsRulovs replied:
Done

pjanck reviewed (changes requested):
Still some opened comments.

Commit:
Corrected double-s initialization(Added a floating point number)

Commit:
Updated IfcProfileDef(Removed "AddAsymmetricIShapecoordinates",
Replaced "AddCircleCoordinates" with "addArc",
removed unneeded spaces,
new function addArc_or_push.back(have to clarify),
Doxycomments to "AddRectangleCoordintaes" (InProgress)).

Commit:
Merge remote-tracking branch 'tumcms/development' into IfcRectangleProfileDef

Commit:
Removed Function GetAsymmetricIShapeCoordinates

hechth commented on Core/src/IfcGeometryConverter/ProfileConverter.h:
```diff
@@ -681,15 +637,12 @@ namespace OpenInfraPlatform {
 						outer_loop.push_back(carve::geom::VECTOR(b * 0.5, (-h * 0.5 + tf)));
 
-						if(r != 0) {
+						if(r != 0.0) {
```
Don't compare doubles using normal == or != operator. See https://stackoverflow.com/questions/12278523/comparing-double-values-in-c/35252979 and http://c-faq.com/fp/fpequal.html

No need to change that here now, just for future implementations!
We should also maybe provide that functionality via BlueFramework or some other utility.

pjanck replied:
See #148 .

hechth commented on Core/src/IfcGeometryConverter/ProfileConverter.h:
```diff
@@ -942,27 +866,27 @@ namespace OpenInfraPlatform {
 					const double tf = t_shape->FlangeThickness * length_factor;
-
-					double r1 = 0;
+					//Redo
+					double r1 = 0.0;
 					if(t_shape->FilletRadius) {
 						r1 = t_shape->FilletRadius * length_factor;
 					}
```
This can still be refactored to `double r1 = t_shape->FilletRadius.value_or(0.0) * length_factor;`

SamuilsRulovs replied:
Done

hechth commented on Core/src/IfcGeometryConverter/ProfileConverter.h:
```diff
@@ -942,27 +866,27 @@ namespace OpenInfraPlatform {
 					}
-
+					//Redo
 					double r2 = 0;
 					if(t_shape->FlangeEdgeRadius) {
 						r2 = t_shape->FlangeEdgeRadius * length_factor;
 					}
```
This can also still be refactored using `value_or(...)`.

SamuilsRulovs replied:
Done

hechth commented on Core/src/IfcGeometryConverter/ProfileConverter.h:
```diff
@@ -942,27 +866,27 @@ namespace OpenInfraPlatform {
 						r2 = t_shape->FlangeEdgeRadius * length_factor;
 					}
-
+					//Redo
 					double r3 = 0;
 					if(t_shape->WebEdgeRadius) {
 						r3 = t_shape->WebEdgeRadius * length_factor;
 					}
```
This can also still be refactored using `value_or(...)`.

SamuilsRulovs replied:
Done

hechth commented on Core/src/IfcGeometryConverter/ProfileConverter.h:
```diff
@@ -942,27 +866,27 @@ namespace OpenInfraPlatform {
 						fs = t_shape->FlangeSlope * angle_factor;
 					}
-
+					//Redo
 					double ws = 0;
 					if(t_shape->WebSlope) {
 						ws = t_shape->WebSlope * angle_factor;
```
This can also still be refactored using `value_or(...)`.

SamuilsRulovs replied:
Done

hechth commented on Core/src/IfcGeometryConverter/ProfileConverter.h:
```diff
@@ -942,27 +866,27 @@ namespace OpenInfraPlatform {
 					}
 					double fs = 0;
-
+					//Redo
 					if(t_shape->FlangeSlope) {
 						fs = t_shape->FlangeSlope * angle_factor;
 					}
```
This can also still be refactored using `value_or(...)`.

SamuilsRulovs replied:
Done

Commit:
Merge remote-tracking branch 'tumcms/development' into IfcRectangleProfileDef

Commit:
Refactored IfcProfileDef using value_or(...), Refactored IfcProfileDef using new function addArc_or_push_back

pjanck commented on Core/src/IfcGeometryConverter/ProfileConverter.h:
```diff
@@ -665,10 +655,15 @@ namespace OpenInfraPlatform {
+								}
+								else {
+									outer_loop.push_back(carve::geom::VECTOR(tw * 0.5, (h * 0.5 - tfTop)));
+								}
```
Perhaps forgot the addARc_or_push_back ?

SamuilsRulovs replied:
Done

pjanck commented on Core/src/IfcGeometryConverter/ProfileConverter.h:
```diff
@@ -834,27 +774,20 @@ namespace OpenInfraPlatform {
 						outer_loop.push_back(carve::geom::VECTOR((b - tw * 0.5), -h * 0.5));
-
-						if(r2 != 0) {
+						//addArc_or_push_back
+						if(r2 != 0.0) {
 							addArc(outer_loop, r2, 0, M_PI_2, b - tw * 0.5 - r2, -h * 0.5 + tf - r2);
 						}
 						else {
 							outer_loop.push_back(carve::geom::VECTOR((b - tw * 0.5), (-h * 0.5 + tf)));
 						}
```
still to do, right?

SamuilsRulovs replied:
Done

pjanck commented on Core/src/IfcGeometryConverter/ProfileConverter.h:
```diff
@@ -1485,33 +1399,17 @@ namespace OpenInfraPlatform {
-					coords.push_back(carve::geom::VECTOR(Webthickness * 0.5, (h * 0.5 - FlangeThickness)));
+					coords.push_back(carve::geom::VECTOR(push_back_x, push_back_y));
 				}
 			}
```
This function seems finished. Doxy-comment, please.

SamuilsRulovs replied:
Done

pjanck reviewed (changes requested):
Great work so far!

Commit:
Added Doxy-comments to finished functions; Replaced remaining conditional statements with new function addArc_or_push_back

Commit:
Added new function "SearchExistingVertex".

pjanck commented on Core/src/IfcGeometryConverter/ProfileConverter.h:
```diff
@@ -1393,6 +1331,18 @@ namespace OpenInfraPlatform {
+				}
+				else {
+					index = polygon->addVertex(position);
+					polygonIndices[ID.str()] = index;
```
Does this compile? `polygonIndices` is not defined in scope ...

SamuilsRulovs replied:
However it compiles :)

SamuilsRulovs replied:
Resolved: 1b6fe62

pjanck replied:
This means that the calling function is not called anywhere. Need to dig deeper into that.

pjanck commented on Core/src/IfcGeometryConverter/ProfileConverter.h:
```diff
@@ -1393,6 +1331,18 @@ namespace OpenInfraPlatform {
 				}
 			}
 
+			void SearchExistingVertex(std::stringstream ID, uint32_t index, carve::geom::vector<3> position) const
```
`index` needs to be passed per reference `&`, otherwise the value will not be available outside the function (within the calling scope).

SamuilsRulovs replied:
Resolved: 1b6fe62

pjanck commented on Core/src/IfcGeometryConverter/ProfileConverter.h:
```diff
@@ -1393,6 +1331,18 @@ namespace OpenInfraPlatform {
 				}
 			}
 
+			void SearchExistingVertex(std::stringstream ID, uint32_t index, carve::geom::vector<3> position) const
```
Also, please pass `std::string` instead of `std::stringstream` and than use the `ID.str()` when calling the function.

SamuilsRulovs replied:
Resolved: 1b6fe62

pjanck commented on Core/src/IfcGeometryConverter/ProfileConverter.h:
```diff
@@ -1393,6 +1331,18 @@ namespace OpenInfraPlatform {
+					index = itFound->second;
+				}
+				else {
+					index = polygon->addVertex(position);
```
Does this compile? `polygon` is not defined ...

SamuilsRulovs replied:
Resolved: 1b6fe62

pjanck reviewed (changes requested):
Really good idea for the function. Just make sure to pass it correct arguments, and to be sure that it compiles.

Commit:
Merge branch 'development' into IfcRectangleProfileDef

Commit:
Corrected "SearchExistingVertex"

Commit:
Added Doxy-comments to "SearchExistingVertex"

Commit:
Corrected "SearchExistingVertex"

Commit:
New function "convertIfcOpenCrossProfileDef" (#TODO)

Commit:
Replaced "std::shared_ptr" with "ExpressReference"

Commit:
Extracted function convertIfcRectangleProfileDef

Commit:
Extracted functions convertIfcTrapeziumProfileDef and convertIfcCircleProfileDef

pjanck commented on Core/src/IfcGeometryConverter/ProfileConverter.h:
```diff
@@ -288,8 +288,8 @@ namespace OpenInfraPlatform {
+				
+				EXPRESSReference<typename IfcEntityTypesT::IfcCenterLineProfileDef> center_line_profile_def =
 					std::dynamic_pointer_cast<typename IfcEntityTypesT::IfcCenterLineProfileDef>(profileDef);
 				if(center_line_profile_def) {
```
this can be solved with `.isOfType<..>` and then having a separate function that handles `IfcCenterLineProfileDef` called with `.as<...>`

SamuilsRulovs replied:
Resolved: a838c88

pjanck commented on Core/src/IfcGeometryConverter/ProfileConverter.h:
```diff
@@ -468,7 +468,17 @@ namespace OpenInfraPlatform {
+			void convertIfcOpenCrossProfileDef(const EXPRESSReference<typename IfcEntityTypesT::IfcOpenCrossProfileDef>& profileDef,
+				std::vector<std::vector<carve::geom::vector<2>>>& paths) const
+			{	
+				throw oip::UnhandledException(convertIfcOpenCrossProfileDef);
```
```suggestion
				throw oip::UnhandledException(profileDef);
```

The argument to UnhandledException MUST be a EXPRESSReference.

SamuilsRulovs replied:
Resolved: 1182a6e

pjanck commented on Core/src/IfcGeometryConverter/ProfileConverter.h:
```diff
@@ -486,97 +496,25 @@ namespace OpenInfraPlatform {
 				// (1/10) IfcRectangleProfileDef SUBTYPE OF IfcParametrizedProfileDef
-				std::shared_ptr<typename IfcEntityTypesT::IfcRectangleProfileDef> rectangle_profile = std::dynamic_pointer_cast<typename IfcEntityTypesT::IfcRectangleProfileDef>(profileDef);		
+				std::shared_ptr<typename IfcEntityTypesT::IfcRectangleProfileDef> rectangle_profile = std::dynamic_pointer_cast<typename IfcEntityTypesT::IfcRectangleProfileDef>(profileDef);
 				if(rectangle_profile) {
```
Same comment as before - `.isOfType` together with `.as`

SamuilsRulovs replied:
Resolved: e76c77e

Commit:
Extracted functions "convertIfcEllipseProfileDef", "convertIfcIShapeProfileDef", "convertIfcLShapeProfileDef" and "convertIfcUShapeProfileDef"

Commit:
Merge remote-tracking branch 'tumcms/development' into IfcRectangleProfileDef

Commit:
Extracted functions "convertIfcCShapeProfileDef", "convertIfcZShapeProfileDef" and "convertIfcTShapeProfileDef"

pjanck commented on Core/src/IfcGeometryConverter/ProfileConverter.h:
```diff
@@ -517,499 +496,361 @@ namespace OpenInfraPlatform {
-						paths.push_back(outer_loop);
-					}
-					return;
+				EXPRESSReference<typename IfcEntityTypesT::IfcTrapeziumProfileDef> trapezium = std::dynamic_pointer_cast<typename IfcEntityTypesT::IfcTrapeziumProfileDef>(profileDef);
+				if (trapezium) {
+					convertIfcTrapeziumProfileDef(EXPRESSReference<typename IfcEntityTypesT::IfcTrapeziumProfileDef>(trapezium), paths, outer_loop);
```
```suggestion
				if (profileDef.isOfType<typename IfcEntityTypesT::IfcTrapeziumProfileDef>()) {
					convertIfcTrapeziumProfileDef(profileDef.as<typename IfcEntityTypesT::IfcTrapeziumProfileDef>(), paths, outer_loop);
```
And similarly everywhere else.

SamuilsRulovs replied:
Resolved:  e76c77e and cdad480

pjanck reviewed (changes requested):
The `if` cases need to be changed accordingly (using `isOfType<..>()` and `as<..>()` functions). See a coment I left at one place.

Commit:
Merge remote-tracking branch 'tumcms/development' into IfcRectangleProfileDef

Commit:
Replaced "shared_ptr" with "EXPRESSReference" in "convertIfcParametrizedProfileDef"

Commit:
Updated IfcParametrizedProfileDef sub-functions

Commit:
Added/Edited "UnhandledException"

Commit:
Replaced "shared_ptr" with "EXPRESSReference" in "ProfileConverter"

SamuilsRulovs commented:
I am going to replace "shared_ptr" with "EXPRESSReference" in function "computeProfile" and "convertIfcArbitraryClosedProfileDef". 
I am going to add Doxy-comments to extracted functions

Commit:
Added Doxy-comments to extracted functions in "IfcParametrizedProfileDef", Moved declaration of "outer_loop" into extracted functions

Commit:
Replaced "shared_ptr" with "EXPRESSReference" in "computeProfile"

Commit:
Adde UnhandledException to "computeProfile" and "convertIfcCompositeProfileDef"

Commit:
Extracted SUBTYPES from "IfcRectangleProfileDef" and "IfcCircleProfileDef", Edited Doxy-comments for functions "convertIfcRectangleProfileDef", "convertIfcCircleProfileDef" and "convertIfcTrapeziumProfileDef"

Commit:
Returned Subtype to "IfcCircleProfileDef", Extracted "StandartRectangle" from "IfcRectangleProfileDef", Edited Doxy-comments by remaining functions

pjanck commented on Core/src/IfcGeometryConverter/ProfileConverter.h:
```diff
@@ -475,143 +469,201 @@ namespace OpenInfraPlatform {
-				if (rectangle_profile) {
-					convertIfcRectangleProfileDef(EXPRESSReference<typename IfcEntityTypesT::IfcRectangleProfileDef>(rectangle_profile), paths, outer_loop);
+				if (profileDef.isOfType<typename IfcEntityTypesT::IfcRectangleProfileDef>()) {
+					convertIfcRectangleProfileDef(profileDef.as<typename IfcEntityTypesT::IfcRectangleProfileDef>(), paths);
```
Since there is no `return` in these `if` clauses, the `throw` at the end of function will always happen. Include returns.

SamuilsRulovs replied:
Resolved: 8838897

pjanck commented on Core/src/IfcGeometryConverter/ProfileConverter.h:
```diff
@@ -475,143 +469,201 @@ namespace OpenInfraPlatform {
 			void convertIfcTrapeziumProfileDef(const EXPRESSReference<typename IfcEntityTypesT::IfcTrapeziumProfileDef>& trapezium,
-				std::vector<std::vector<carve::geom::vector<2>>>& paths, std::vector<carve::geom::vector<2>> outer_loop) const {
+				std::vector<std::vector<carve::geom::vector<2>>>& paths) const {
+				// **************************************************************************************************************************
+				// ENTITY IfcTrapeziumProfileDef
+				// https://standards.buildingsmart.org/IFC/DEV/IFC4_2/FINAL/HTML/schema/ifcprofileresource/lexical/ifctrapeziumprofiledef.htm
+				// SUBTYPE OF(IfcParametrizedProfileDef);
+				// END_ENTITY;
+				// **************************************************************************************************************************
```
The ENTITY definitions are not correct. Please see the Schema file and include the correct definitions.

SamuilsRulovs replied:
Resolved:  8838897

pjanck reviewed (changes requested):
Two comments that apply for the whole file, not only where they are posted.

Commit:
added "return" to if clauses, edited ENTITY definitions
